Introduce xpc_string_get_wtfstring() for bounds safety and efficiency

ddkilzer opened the pull request:
#### a33eba312fbbd5b863c3c65504b3bef9fa659dd8
<pre>
Introduce xpc_string_get_wtfstring() for bounds safety and efficiency
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=288750">https://bugs.webkit.org/show_bug.cgi?id=288750</a>&gt;
&lt;<a href="https://rdar.apple.com/145785145">rdar://145785145</a>&gt;

Reviewed by Geoffrey Garen.

* Source/WTF/wtf/spi/darwin/XPCSPI.h:
(xpc_string_get_length):
- Add declaration for open source builds.
(xpc_string_get_wtfstring): Add.
- Convenience method for converting C-string returned from
  xpc_string_get_string_ptr() into a WTF::String.  This is more
  efficient since it creates a span with the known length of the
  C-string to construct the WTF::String.
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
- Make use of xpc_string_get_wtfstring().
- Add a `NOLINT` comment to avoid check-webkit-style warnings for the
  second use of xpc_string_get_string_ptr() which doesn&apos;t need a
  WTF::String.
* Tools/Scripts/webkitpy/style/checkers/cpp.py:
(check_safer_cpp):
- Add check for xpc_string_get_string_ptr().
(CppChecker):
- Enable &apos;safercpp/xpc_string_get_string_ptr&apos; checker by default.
* Tools/Scripts/webkitpy/style/checkers/cpp_unittest.py:
(WebKitStyleTest.test_safer_cpp):
- Add test for &apos;safercpp/xpc_string_get_string_ptr&apos; checker.

Canonical link: <a href="https://commits.webkit.org/291513@main">https://commits.webkit.org/291513@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b3f1abb86edcb75152942c262b471c803dab45a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92547 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1701 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97532 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43054 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94597 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12370 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20550 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70889 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28336 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95549 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9379 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83790 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51221 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/92042 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9075 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1428 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42385 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/85257 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79393 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1369 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99558 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91213 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19598 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14452 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79895 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19848 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79672 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79177 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23710 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/999 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12610 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14893 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19582 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24754 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113861 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19269 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32913 "Found 316 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/function-apply-many-args.js.layout-no-llint, microbenchmarks/array-from-object.js.lockdown, microbenchmarks/array-prototype-with-storage.js.bytecode-cache, microbenchmarks/array-prototype-with-storage.js.default, microbenchmarks/array-prototype-with-storage.js.dfg-eager, microbenchmarks/array-prototype-with-storage.js.dfg-eager-no-cjit-validate, microbenchmarks/array-prototype-with-storage.js.eager-jettison-no-cjit, microbenchmarks/array-prototype-with-storage.js.lockdown, microbenchmarks/array-prototype-with-storage.js.mini-mode, microbenchmarks/array-prototype-with-storage.js.no-cjit-collect-continuously ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22729 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21009 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
```diff
@@ -168,8 +168,8 @@ void XPCServiceEventHandler(xpc_connection_t peer)
                 @autoreleasepool {
-                    xpc_array_apply(languages, makeBlockPtr([&newLanguages](size_t index, xpc_object_t value) {
-                        newLanguages.append(String::fromUTF8(xpc_string_get_string_ptr(value)));
+                    xpc_array_apply(languages.get(), makeBlockPtr([&newLanguages](size_t index, xpc_object_t value) {
```
`.get()`?

ddkilzer replied:
Yep, fixed.

aproskuryakov commented on Source/WTF/wtf/spi/darwin/XPCSPI.h:
```diff
@@ -275,3 +276,8 @@ inline String xpc_dictionary_get_wtfstring(xpc_object_t xdict, ASCIILiteral key)
     return String::fromUTF8(cstring);
 }
+
+inline String xpc_string_get_wtfstring(xpc_object_t xvalue)
```
This seems like the wrong name and place for this function. It's not an SPI, so it should be in a normal place, and named according to WebKit style rules.

geoffreygaren replied:
I’m guessing Dave’s following the example of xpc_dictionary_get_data_span and xpc_dictionary_get_wtfstring. 

We want to associate functions like this with the API they came from because their purpose is to add missing memory safety facts that are properties of  the source API. 

Ideally code like this should be at the outermost edges of the project, to emphasize the fact that, inside WebKit, programmers should not manually assign bounds to buffers. 

Would it be better if we named this file XPCSPIAndAdditions.h?

ddkilzer replied:
Yes, I'm following the example from a69cdc2b13007677cbada1e6d74ca8009a474af1.

aproskuryakov replied:
I agree with Geoff that we should be further separating unsafe SPI functions that people shouldn't be using from safe wrappers. So why rename when we can add a new header file?

ddkilzer replied:
Can you file a new Bugzilla for the requested changes?  I don't think that needs to block landing this fix.

aproskuryakov replied:
How is this suddenly a task for me? I'm just pointing out shoddy work here, that's all.

geoffreygaren replied:
I don’t think we need to call an honest engagement with feedback ‘shoddy’. 

For my part, I’m still trying to understand what the proposal is. Is the proposal a new file named XPCAdditions.h, which contains three functions?

aproskuryakov replied:
Again, it's obviously shoddy coding to put non-SPI code into SPI headers. Please stop pretending that it's my job to solve it for you just because I pointed it out.

ddkilzer replied:
When I first saw `xpc_dictionary_get_wtfstring()`, I was also surprised at its naming convention and header location, but I assumed there was some change in direction for helper functions (perhaps to make them easier to find with autocomplete?), so I went with the flow.

> How is this suddenly a task for me? I'm just pointing out shoddy work here, that's all.

I think it's reasonable to ask someone to file a bug report if they feel so moved as to point it out.

In the interest of moving this discussion forward—maybe even resolving it—I took the time to file a bug:

[Bug 288984](https://bugs.webkit.org/show_bug.cgi?id=288984):  Move WebKit utility functions out of XPCSPI.h and rename

ddkilzer replied:
Initial PR:  https://github.com/WebKit/WebKit/pull/41806

aproskuryakov replied:
Thank you, it looks nice!

johnwilander replied:
> Again, it's obviously shoddy coding to put non-SPI code into SPI headers. Please stop pretending that it's my job to solve it for you just because I pointed it out.

Could you avoid using judgmental language such as "obviously shoddy coding," please? I don't think it helps code quality or collaboration.

Shoddy: badly made or done; lacking moral principle; sordid

geoffreygaren reviewed (approved):
I’m going to r+ this patch because it advances our Safer CPP goals and meets project guidelines; and I think we should continue the discussion about the best way to organize our headers.

Commit:
Introduce xpc_string_get_wtfstring() for bounds safety and efficiency
<https://bugs.webkit.org/show_bug.cgi?id=288750>
<rdar://145785145>

Reviewed by Geoffrey Garen.

* Source/WTF/wtf/spi/darwin/XPCSPI.h:
(xpc_string_get_length):
- Add declaration for open source builds.
(xpc_string_get_wtfstring): Add.
- Convenience method for converting C-string returned from
  xpc_string_get_string_ptr() into a WTF::String.  This is more
  efficient since it creates a span with the known length of the
  C-string to construct the WTF::String.
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
- Make use of xpc_string_get_wtfstring().
- Add a `NOLINT` comment to avoid check-webkit-style warnings for the
  second use of xpc_string_get_string_ptr() which doesn't need a
  WTF::String.
* Tools/Scripts/webkitpy/style/checkers/cpp.py:
(check_safer_cpp):
- Add check for xpc_string_get_string_ptr().
(CppChecker):
- Enable 'safercpp/xpc_string_get_string_ptr' checker by default.
* Tools/Scripts/webkitpy/style/checkers/cpp_unittest.py:
(WebKitStyleTest.test_safer_cpp):
- Add test for 'safercpp/xpc_string_get_string_ptr' checker.

Canonical link: https://commits.webkit.org/291513@main
